File: computer/lib/context-loader.js
/**
 * Context Loader
 *
 * Loads context/knowledge for agents:
 * - Reads knowledge files that define what an agent "knows"
 * - Resolves [[wiki-links]] in knowledge files
 * - Loads files matching include patterns
 * - Respects max_tokens limits
 */

import fs from 'fs/promises';
import path from 'path';
import { glob } from 'glob';

/**
 * Load context for an agent based on its context configuration
 *
 * @param {object} contextConfig - Agent's context configuration
 * @param {string} vaultPath - Base vault path
 * @param {object} options - Options like max_tokens
 * @returns {Promise<{files: Array, totalTokens: number, content: string}>}
 */
export async function loadAgentContext(contextConfig, vaultPath, options = {}) {
  const maxTokens = options.max_tokens || contextConfig.max_tokens || 50000;
  const loadedFiles = [];
  let totalContent = '';
  let estimatedTokens = 0;

  // 1. Load knowledge file if specified
  if (contextConfig.knowledge_file) {
    const knowledgeResult = await loadKnowledgeFile(
      contextConfig.knowledge_file,
      vaultPath,
      maxTokens - estimatedTokens
    );
    loadedFiles.push(...knowledgeResult.files);
    totalContent += knowledgeResult.content;
    estimatedTokens += knowledgeResult.estimatedTokens;
  }

  // 2. Load files matching include patterns
  if (contextConfig.include && Array.isArray(contextConfig.include)) {
    for (const pattern of contextConfig.include) {
      if (estimatedTokens >= maxTokens) break;

      const files = await resolveGlobPattern(pattern, vaultPath);
      for (const file of files) {
        if (estimatedTokens >= maxTokens) break;
        if (loadedFiles.some(f => f.path === file)) continue; // Skip duplicates

        const fileResult = await loadFile(file, vaultPath, maxTokens - estimatedTokens);
        if (fileResult) {
          loadedFiles.push(fileResult);
          totalContent += `\n\n---\n## ${file}\n\n${fileResult.content}`;
          estimatedTokens += fileResult.estimatedTokens;
        }
      }
    }
  }

  return {
    files: loadedFiles,
    totalTokens: estimatedTokens,
    content: totalContent
  };
}

/**
 * Load a knowledge file and resolve its [[wiki-links]]
 *
 * @param {string} knowledgePath - Path to the knowledge file
 * @param {string} vaultPath - Base vault path
 * @param {number} remainingTokens - Token budget remaining
 * @returns {Promise<{files: Array, content: string, estimatedTokens: number}>}
 */
async function loadKnowledgeFile(knowledgePath, vaultPath, remainingTokens) {
  const fullPath = path.join(vaultPath, knowledgePath);
  const loadedFiles = [];
  let content = '';
  let estimatedTokens = 0;

  try {
    const knowledgeContent = await fs.readFile(fullPath, 'utf-8');

    // Add the knowledge file itself as context header
    content += `## Knowledge: ${knowledgePath}\n\n${knowledgeContent}\n`;
    estimatedTokens += estimateTokens(knowledgeContent);
    loadedFiles.push({
      path: knowledgePath,
      type: 'knowledge',
      content: knowledgeContent,
      estimatedTokens: estimateTokens(knowledgeContent)
    });

    // Extract and resolve [[wiki-links]]
    const wikiLinks = extractWikiLinks(knowledgeContent);

    for (const link of wikiLinks) {
      if (estimatedTokens >= remainingTokens) break;

      const resolvedPath = resolveWikiLink(link, vaultPath);
      if (!resolvedPath) continue;
      if (loadedFiles.some(f => f.path === resolvedPath)) continue;

      const fileResult = await loadFile(resolvedPath, vaultPath, remainingTokens - estimatedTokens);
      if (fileResult) {
        loadedFiles.push(fileResult);
        content += `\n\n---\n## ${resolvedPath}\n\n${fileResult.content}`;
        estimatedTokens += fileResult.estimatedTokens;
      }
    }

  } catch (e) {
    console.warn(`[ContextLoader] Could not load knowledge file ${knowledgePath}:`, e.message);
  }

  return { files: loadedFiles, content, estimatedTokens };
}

/**
 * Extract [[wiki-links]] from content
 *
 * Supports:
 * - [[simple-link]]
 * - [[link|display text]]
 * - [[folder/nested-link]]
 *
 * @param {string} content
 * @returns {string[]} Array of link targets (without [[ ]])
 */
function extractWikiLinks(content) {
  const linkRegex = /\[\[([^\]|]+)(?:\|[^\]]+)?\]\]/g;
  const links = [];
  let match;

  while ((match = linkRegex.exec(content)) !== null) {
    links.push(match[1].trim());
  }

  return [...new Set(links)]; // Remove duplicates
}

/**
 * Resolve a wiki-link to an actual file path
 *
 * @param {string} link - The wiki link target
 * @param {string} vaultPath - Base vault path
 * @returns {string|null} Resolved path or null if not found
 */
function resolveWikiLink(link, vaultPath) {
  // Add .md if not present
  const linkWithExt = link.endsWith('.md') ? link : `${link}.md`;

  // Try exact path first
  const exactPath = path.join(vaultPath, linkWithExt);
  try {
    // Use sync check for simplicity (will be called rarely)
    const stats = require('fs').statSync(exactPath);
    if (stats.isFile()) {
      return linkWithExt;
    }
  } catch {
    // File doesn't exist at exact path
  }

  // Could add fuzzy matching here later (search for file by name in vault)
  return null;
}

/**
 * Resolve a glob pattern to actual file paths
 *
 * @param {string} pattern - Glob pattern
 * @param {string} vaultPath - Base vault path
 * @returns {Promise<string[]>} Array of matching file paths (relative to vault)
 */
async function resolveGlobPattern(pattern, vaultPath) {
  try {
    const matches = await glob(pattern, {
      cwd: vaultPath,
      nodir: true,
      ignore: ['**/node_modules/**', '**/.obsidian/**']
    });
    return matches;
  } catch (e) {
    console.warn(`[ContextLoader] Error resolving pattern ${pattern}:`, e.message);
    return [];
  }
}

/**
 * Load a single file
 *
 * @param {string} filePath - Relative path to file
 * @param {string} vaultPath - Base vault path
 * @param {number} maxTokens - Max tokens to include
 * @returns {Promise<{path: string, content: string, estimatedTokens: number, truncated: boolean}|null>}
 */
async function loadFile(filePath, vaultPath, maxTokens) {
  const fullPath = path.join(vaultPath, filePath);

  try {
    let content = await fs.readFile(fullPath, 'utf-8');
    let truncated = false;
    let tokens = estimateTokens(content);

    // Truncate if necessary
    if (tokens > maxTokens) {
      const ratio = maxTokens / tokens;
      const charLimit = Math.floor(content.length * ratio * 0.9); // 90% to be safe
      content = content.substring(0, charLimit) + '\n\n[... truncated ...]';
      tokens = maxTokens;
      truncated = true;
    }

    return {
      path: filePath,
      type: 'file',
      content,
      estimatedTokens: tokens,
      truncated
    };

  } catch (e) {
    console.warn(`[ContextLoader] Could not load file ${filePath}:`, e.message);
    return null;
  }
}

/**
 * Rough token estimation (~4 chars per token for English text)
 *
 * @param {string} text
 * @returns {number}
 */
function estimateTokens(text) {
  return Math.ceil(text.length / 4);
}

/**
 * Format loaded context as a system prompt section
 *
 * @param {object} contextResult - Result from loadAgentContext
 * @returns {string}
 */
export function formatContextForPrompt(contextResult) {
  if (!contextResult.content || contextResult.files.length === 0) {
    return '';
  }

  let prompt = `\n\n## Project Knowledge\n`;
  prompt += `The following context has been loaded for your reference (${contextResult.files.length} files, ~${contextResult.totalTokens} tokens):\n`;
  prompt += contextResult.content;

  return prompt;
}

export default {
  loadAgentContext,
  formatContextForPrompt,
  extractWikiLinks
};

File: parachute/lib/context_loader.py
"""
Context/knowledge file loader for agents.

Loads context files from the vault and formats them for inclusion in prompts.
"""

import logging
from pathlib import Path
from typing import Any, Optional

import frontmatter

logger = logging.getLogger(__name__)

# Rough estimate: 4 chars per token
CHARS_PER_TOKEN = 4


async def load_agent_context(
    context_config: dict[str, Any],
    vault_path: Path,
    options: Optional[dict[str, Any]] = None,
) -> dict[str, Any]:
    """
    Load context files based on agent context configuration.

    Args:
        context_config: Agent's context configuration with 'include', 'exclude', etc.
        vault_path: Path to the vault
        options: Additional options like max_tokens

    Returns:
        Dict with 'content', 'files', 'totalTokens', 'truncated'
    """
    options = options or {}
    max_tokens = context_config.get("max_tokens", options.get("max_tokens", 50000))
    max_chars = max_tokens * CHARS_PER_TOKEN

    include_patterns = context_config.get("include", [])
    exclude_patterns = context_config.get("exclude", [])
    knowledge_file = context_config.get("knowledge_file")

    # Collect files to load
    files_to_load: list[Path] = []

    # Handle knowledge_file (single file path)
    if knowledge_file:
        kf_path = vault_path / knowledge_file
        if kf_path.exists():
            files_to_load.append(kf_path)

    # Handle include patterns
    for pattern in include_patterns:
        if "*" in pattern:
            # Glob pattern
            for match in vault_path.glob(pattern):
                if match.is_file() and not _matches_exclude(match, exclude_patterns, vault_path):
                    files_to_load.append(match)
        else:
            # Direct path
            file_path = vault_path / pattern
            if file_path.exists() and file_path.is_file():
                files_to_load.append(file_path)

    # Load content from files
    loaded_files: list[str] = []
    content_parts: list[str] = []
    total_chars = 0
    truncated = False

    for file_path in files_to_load:
        if total_chars >= max_chars:
            truncated = True
            break

        try:
            relative_path = str(file_path.relative_to(vault_path))
            file_content = await _load_file_content(file_path)

            if file_content:
                remaining = max_chars - total_chars
                if len(file_content) > remaining:
                    file_content = file_content[:remaining] + "\n[...truncated]"
                    truncated = True

                content_parts.append(f"## {relative_path}\n\n{file_content}")
                loaded_files.append(relative_path)
                total_chars += len(file_content)

        except Exception as e:
            logger.warning(f"Error loading context file {file_path}: {e}")

    combined_content = "\n\n---\n\n".join(content_parts)
    estimated_tokens = total_chars // CHARS_PER_TOKEN

    return {
        "content": combined_content,
        "files": loaded_files,
        "totalTokens": estimated_tokens,
        "truncated": truncated,
    }


def _matches_exclude(file_path: Path, patterns: list[str], vault_path: Path) -> bool:
    """Check if a file matches any exclude pattern."""
    from parachute.lib.vault_utils import matches_patterns

    relative = str(file_path.relative_to(vault_path))
    return matches_patterns(relative, patterns)


async def _load_file_content(file_path: Path) -> Optional[str]:
    """Load content from a file, parsing frontmatter if markdown."""
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()

        # Parse frontmatter for markdown files
        if file_path.suffix == ".md":
            post = frontmatter.loads(content)
            return post.content.strip()

        return content.strip()
    except Exception as e:
        logger.debug(f"Failed to read {file_path}: {e}")
        return None


def format_context_for_prompt(context_result: dict[str, Any]) -> str:
    """
    Format loaded context for inclusion in a system prompt.

    Args:
        context_result: Result from load_agent_context

    Returns:
        Formatted string for system prompt
    """
    if not context_result.get("content"):
        return ""

    header = f"\n\n## Project Knowledge\n"
    header += f"The following context has been loaded for your reference "
    header += f"({len(context_result['files'])} files, ~{context_result['totalTokens']} tokens):\n\n"

    footer = ""
    if context_result.get("truncated"):
        footer = "\n\n*[Context truncated due to token limit]*"

    return header + context_result["content"] + footer
